feat/move-entity-to-space (GEO-331)

jonathanprozzi opened the pull request:
## Overview

- Linear issue: GEO-331

This PR adds the ability for users to move an entity from one space to another (the user must be an editor in each space). If a user is an editor, they will see an option for "Move to Space" in the Entity menu for the first space. They can then choose a second space to move the Entity to. 

Selecting the second space opens the MoveEntity review UI, which shows the user the Entity details and the spaces (to and from). 

There are 2 transactions involved in this process:
- Create the new triples for the entity in the new space
- Delete the triples in the first space
- Each transaction state is monitored in the separate publish flows
- It's possible for the first tx to succeed and the second to fail -- these are each separate proposals so if this occurs, editors in the first space can choose to revert (each of these flows is a separate proposal that is generated as part of the publish flow)
- Once both succeed, there is a 2 second pause to allow the user to see that both flows have succeeded. After 2 seconds, the UI closes
- This also moves a part of the logic from the publish flow to the Review component (and adds a resolve to the promise)

## Questions/Notes

- This worked while testing with a local development environment, but I ran into issues with the IPFS staging Graph being down (503 error)
- Do we want to redirect to the new entity in the new space? @baiirun I believe we discussed this initially
- When testing locally, the transactions occur very quickly and there is an occasional 'pending' status (particularly in MetaMask) but the transactions fully succeed and resolve. This may not be the case on mainnet, but we should keep an eye on this for user feedback.

@baiirun going to mark this ready for review -- I'm sure that there are areas that will need to be addressed as part of review, but it's ready to look at! 

`lint` and `build` both succeed locally.

Demo with local environment (some small inconsistencies with this such as no space images and the entity name has been acting strangely since needing to change away from IPFS staging URL that is set in `environment.ts`
![move-entity-local-dev-demo](https://github.com/geobrowser/geogenesis/assets/9438776/bf24a0b7-226d-410a-a361-0f5f160cee72)

baiirun commented:
I'm not sure if that left to right (->) arrow makes sense with the Create Triples on the left hand side

baiirun commented on apps/web/partials/move-entity/move-entity-review.tsx:
```diff
@@ -0,0 +1,288 @@
+        ...t,
+        type: 'deleteTriple',
+      }));
+      deleteActions.map(action => remove(action));
```
forEach instead of map

jonathanprozzi commented on apps/web/partials/move-entity/move-entity-review.tsx:
```diff
@@ -0,0 +1,288 @@
+
+import { MoveEntityReviewPage } from './move-entity-review-page';
+
+// added the overflow classes up here to make sure mobile devices can scroll for longer content in the Entity Review UI
```
Great call! moved

jonathanprozzi commented on apps/web/design-system/icon.tsx:
```diff
@@ -96,7 +100,11 @@ export type IconName =
   | 'disconnectWallet'
   | 'bulkEdit'
-  | 'member';
+  | 'member'
+  | 'moveSpace'
+  | 'checkCircleSmall'
+  | 'checkCircle'
+  | 'checkClose';
```
That makes a ton of sense. I noticed that we sometimes didn't use it. Given that it's importing all of those unused icons I'll definitely import them directly

jonathanprozzi commented:
> 

I'm wondering if maybe it isn't necessary now that we explicitly have "Step 1 -" and "Step 2 -"

jonathanprozzi commented on apps/web/partials/move-entity/move-entity-review-page.tsx:
```diff
@@ -0,0 +1,90 @@
+          <ImageZoom
+            key={`image-${triple.attributeId}-${triple.value.id}-${triple.id}`}
+            imageSrc={triple.value.value}
+          />
+        );
+      case 'date':
+        return <DateField isEditing={false} value={triple.value.value} />;
+      case 'url':
+        return <WebUrlField isEditing={false} value={triple.value.value} />;
+      case 'entity': {
+        return (
+          <div key={`entity-${triple.attributeId}-${triple.value.id}-${triple.id}`} className="mt-1">
+            <LinkableChip href={NavUtils.toEntity(triple.space, triple.value.id)}>
+              {triple.value.name || triple.value.id}
+            </LinkableChip>
+          </div>
+        );
+      }
+      case 'number':
+        return null;
+    }
+  };
+
+  return (
+    <>
+      {Object.entries(groupedTriples).map(([attributeId, triples], index) => {
+        if (attributeId === SYSTEM_IDS.BLOCKS) return null;
+
+        return (
+          <div key={`${entityId}-${attributeId}-${index}`} className="break-words">
+            <Text as="p" variant="bodySemibold">
+              {triples[0].attributeName || attributeId}
+            </Text>
+            <div className="flex flex-wrap gap-2">{triples.map(tripleToEditableField)}</div>
+          </div>
+        );
+      })}
+    </>
+  );
+}
```
Good question. I basically reused the component from the entity page and barely changed it, but also wondering if there's ever a case where there'll be differences in how it's used contextually anyway like you mentioned. 

One potential benefit of having it in a shared partial would be if we added additional Blocks that would have specific renders. That would be something we'd need to add to every instance instead of just the shared partial

baiirun replied:
I'd say we leave it for now instead of trying to optimize too hard

jonathanprozzi commented on apps/web/partials/move-entity/move-entity-review.tsx:
```diff
@@ -0,0 +1,288 @@
+}
+
+function ProgressBar({ txState }: { txState: ReviewState }) {
+  const getBgClassByState = (index: number, state: ReviewState) => {
+    let stateValue = 0;
+    switch (state) {
+      case 'idle':
+        stateValue = 0;
+        break;
+      case 'publishing-ipfs':
+        stateValue = 1;
+        break;
+      case 'signing-wallet':
+        stateValue = 2;
+        break;
+      case 'publishing-contract':
+        stateValue = 3;
+        break;
+      case 'publish-complete':
+        stateValue = 4;
+        break;
+      default:
+        stateValue = 0;
+        break;
+    }
+    return state === 'idle' ? 'bg-grey-02' : index <= stateValue ? 'bg-text' : 'bg-grey-02'; // handle the idle case and then the rest
+  };
```
Ahh. Great point about it being needlessly recreated.

baiirun commented on apps/web/partials/move-entity/move-entity-review.tsx:
```diff
@@ -0,0 +1,320 @@
+    const onCreateNewTriples = (): CreateTripleAction[] => {
+      return triples.map(t => ({
+        ...t,
+        entityName: entityNameFromTriples,
```
This _should_ automatically be populated by the existing `entityName` on each triple.

jonathanprozzi replied:
It definitely should -- I added that in since I was seeing some that had a null name, but going to remove this since it should be working as intended and there may be an issue elsewhere where the entityName isnt persisting

baiirun commented on apps/web/partials/move-entity/move-entity-review.tsx:
```diff
@@ -0,0 +1,320 @@
+    }
+    // close the review UI after displaying the state messages for 2 seconds
+    await sleepWithCallback(() => {
+      router.push(`/space/${spaceIdTo}`);
```
At this point I think is where we do 

```tsx
batch(() => {
  deleteActions.forEach(t => remove(t))
  createActions.forEach(t => create(t))
})
```

jonathanprozzi commented:
Fixed the redirect and removed the unnecessary batch calls for create and remove since that's handled in the makeProposal flow now:

![move-entity-flow](https://github.com/geobrowser/geogenesis/assets/9438776/ed3ed1d1-5926-4cbf-b652-588682f6a6ad)

jonathanprozzi commented:
Flow showing when a user rejects a tx:

- Rejecting the first tx -- hitting 'Publish and Move' restarts from the beginning
- Rejecting the second tx after a successful first -- hitting 'Publish and Move' starts at the delete flow

![move-entity-rejected-flow-1](https://github.com/geobrowser/geogenesis/assets/9438776/724b716c-ca73-4cc2-9f85-377f6bf69372)

![move-entity-forced-error-1](https://github.com/geobrowser/geogenesis/assets/9438776/bbdb9602-35b3-4ca1-93ef-d221b963bc76)

Some other edge cases for consideration:

- First tx succeeds but user closes out the review UI or closes the tab -> create proposal has passed but the delete isn't, and the Publish and Move would _re-create_ the triples in the new space. There are a few ways we could handle this but also has UX considerations
- Change 'Re-try' to 'Copy error' if it's not a user rejection -> If it's a full error, the 'Copy error' could be more useful. The user could resume publication after step 1 by clicking the 'Publish and Move' instead of 'Re-try' 
@baiirun and I were discussing some of the tradeoffs for these edge cases

baiirun reviewed (approved):
🚀
